web/packages/teleterm/src/services/tshd/grpcContextBridgeClient.ts

```diff
@@ -0,0 +1,198 @@
+  return {
+    send: (...args) => original.send(...args),
+    complete: (...args) => original.complete(...args),
+  };
+}
+
+function objectifyResponses<O extends object>(
+  original: RpcOutputStream<O>
+): RpcOutputStream<O> {
+  return {
+    ...original,
+    onMessage: (...args) => original.onMessage(...args),
+    onError: errorCallback =>
+      original.onError(e => errorCallback(objectifyError(e) as Error)),
+    onComplete: (...args) => original.onComplete(...args),
+    onNext: (...args) => original.onNext(...args),
+  };
+}
+
+function objectifyError(error: unknown): object {
+  const e = error as RpcError;
+  return {
+    ...e,
+    message: e.message,
+    cause: e.cause,
+    stack: e.stack,
+    code: e.code,
+    isResolvableWithRelogin: e.meta['is-resolvable-with-relogin'] === '1',
+  };
+}
+
+type PublicProperties<T> = { [K in keyof T]: T[K] }; //keyof only sees public properties
+
+function validateAbortSignalType(
+  abortSignal: AbortSignal | ObjectifiedAbortSignal
+) {
+  if (abortSignal && abortSignal['canBePassedThroughContextBridge'] !== true) {
+    throw new Error(
+      'You must not pass AbortSignal instance. Use objectified version (ObjectifedAbortSignal).'
+    );
+  }
+}
+
+async function objectifyPromiseRejection<TResult>(
+  p: Promise<TResult>
+): Promise<TResult> {
+  try {
+    return await p;
+  } catch (e) {
+    throw objectifyError(e);
+  }
+}
+
+function objectifyThenRejection<TResult>(
+  originalThen: Promise<TResult>['then']
+): Promise<TResult>['then'] {
+  return (onFulfilled, onRejected) => {
+    // If onRejected callback is provided, then it will handle the rejection.
+    if (onRejected) {
+      return originalThen(onFulfilled, reason =>
+        onRejected(objectifyError(reason))
+      );
+    }
+    return objectifyPromiseRejection(originalThen(onFulfilled));
+  };
+}
+
+export function getObjectifiedMethodsInterceptor(): RpcInterceptor {
+  return {
+    interceptUnary(next, method, input, options) {
+      validateAbortSignalType(options.abort);
+
+      const output = next(method, input, options);
+      // UnaryCall is a class with a private field,
+      // so we can't assign an object to it.
+      // The workaround is satisfying only the public part of it
+      // and then casting to UnaryCall.
+      const objectifiedUnaryCall: PublicProperties<UnaryCall> = {
+        method: output.method,
+        request: output.request,
+        requestHeaders: output.requestHeaders,
+        status: objectifyPromiseRejection(output.status),
+        trailers: objectifyPromiseRejection(output.trailers),
+        headers: objectifyPromiseRejection(output.headers),
+        response: objectifyPromiseRejection(output.response),
+        // then is a method on UnaryClass,
+        // we have to set correct "this" before passing it further.
+        then: objectifyThenRejection(output.then.bind(output)),
+      };
+      return { ...output, ...objectifiedUnaryCall } as UnaryCall;
```

Comment:
I didn't understand why we spread on `object` too, so I wrote a little helper test to verify whether the returned unary call is cloneable.

<details>
<summary>Cloneable test</summary>

```diff
diff --git a/web/packages/teleterm/src/services/tshd/grpcContextBridgeClient.test.ts b/web/packages/teleterm/src/services/tshd/grpcContextBridgeClient.test.ts
index 9dbe632f9d..f277d2060f 100644
--- a/web/packages/teleterm/src/services/tshd/grpcContextBridgeClient.test.ts
+++ b/web/packages/teleterm/src/services/tshd/grpcContextBridgeClient.test.ts
@@ -1,3 +1,6 @@
+/**
+ * @jest-environment node
+ */
 /**
  * Teleport
  * Copyright (C) 2023  Gravitational, Inc.
@@ -16,7 +19,17 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import { objectifyAbortSignal } from 'teleterm/services/tshd/grpcContextBridgeClient';
+import * as grpc from '@grpc/grpc-js';
+
+import * as apiService from 'gen-proto-ts/teleport/lib/teleterm/v1/service_pb.grpc-server';
+import { TerminalServiceClient } from 'gen-proto-ts/teleport/lib/teleterm/v1/service_pb.client';
+
+import { GrpcTransport } from '@protobuf-ts/grpc-transport';
+
+import {
+  getObjectifiedMethodsInterceptor,
+  objectifyAbortSignal,
+} from 'teleterm/services/tshd/grpcContextBridgeClient';
 
 test('objectified abort signal reads up-to-date signal.aborted', () => {
   const controller = new AbortController();
@@ -25,3 +38,75 @@ test('objectified abort signal reads up-to-date signal.aborted', () => {
   controller.abort();
   expect(objectifyAbortSignal(controller.signal).aborted).toBe(true);
 });
+
+test('unary response is cloneable', async () => {
+  const service: Partial<apiService.ITerminalService> = {
+    updateTshdEventsServerAddress: (call, callback) => {
+      callback(null, {});
+      // TODO: Add a separate test for when an unary response returns an error.
+      // callback(new Error("Whoops"), null);
+    },
+  };
+  const server = new grpc.Server();
+  server.addService(apiService.terminalServiceDefinition, service);
+  const port = await bindServer(server, 'localhost:0');
+  server.start();
+
+  // If port is less than zero than bind failed.
+  expect(port).toBeGreaterThan(0);
+
+  const client = new TerminalServiceClient(
+    new GrpcTransport({
+      host: `localhost:${port}`,
+      channelCredentials: grpc.credentials.createInsecure(),
+      interceptors: [getObjectifiedMethodsInterceptor()],
+    })
+  );
+
+  const unaryCall = client.updateTshdEventsServerAddress({ address: 'foo' });
+
+  for (const propertyName in unaryCall) {
+    let property = unaryCall[propertyName];
+
+    if (property['then']) {
+      property = await property;
+    }
+
+    // Functions cannot be passed to structuredClone, but they can be passed through the context
+    // bridge.
+    if (typeof property === 'function') {
+      console.log(`TODO: Skipping ${propertyName}, add a specific test for it`);
+      continue;
+    }
+
+    expect(() => {
+      try {
+        structuredClone(property);
+      } catch (err) {
+        throw new Error(
+          `Calling structuredClone on the '${propertyName}' property has thrown an error: ${err}`,
+          { cause: err }
+        );
+      }
+    }).not.toThrow();
+  }
+
+  expect(() => structuredClone(unaryCall)).not.toThrow();
+
+  await unaryCall;
+  server.forceShutdown();
+});
+
+const bindServer = (server: grpc.Server, address: string): Promise<number> =>
+  new Promise((resolve, reject) => {
+    server.bindAsync(
+      address,
+      grpc.ServerCredentials.createInsecure(),
+      (error, port) => {
+        if (error) {
+          reject(error);
+        }
+        resolve(port);
+      }
+    );
+  });

```
</details>

It calls `structuredClone` on each property of the unary call and then on the unary call itself. It ignores functions as those cannot be passed to `structuredClone` but can be passed through the context bridge.

In general I just wanted to determine how much of the returned unary call we have to change. The `method` property does not seem to be cloneable. But also it appears that spreading on `output` might not be necessary?

I was hoping this would give me a better understanding of what to do with those promises you mentioned in https://github.com/gravitational/teleport/pull/38573/files#r1512524830. But tbh it didn't get me much closer.

We probably need a separate test to verify that the errors get wrapped properly?

Reply:
Furthermore, I'm not sure if we need to absolutely "objectify" everything and if interceptors are the way to address that.

Take a look at my comment about awaiting the response below. https://github.com/gravitational/teleport/pull/38573#discussion_r1514709835 We could have the "objectified" client with a somewhat stricter API than the generic client provided by the library. One where maybe you cannot inspect `method` and `requestHeaders`, but one that makes the callsite use the correct way to await a response and that wraps the errors properly.

Reply:
>I didn't understand why we spread on object too, so I wrote a little helper test to verify whether the returned unary call is cloneable.

So the spread of `output` is actually not needed, I added it "just in case". Since we assign the objectified call to `PublicProperties<UnaryCall>` then will get a TS error if a new property is added to public API of `UnaryCall`.
I will remove it, since it is just confusing.

>In general I just wanted to determine how much of the returned unary call we have to change. The method property does not seem to be cloneable.

I'm not sure if checking if `structuredClone` works is a correct way to test it. AFAIK context bridge just drops the prototype, because I see `methods` object passing the bridge:
<img width="885" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/9064f833-13d7-475c-b55f-41f5262175e8">

>We probably need a separate test to verify that the errors get wrapped properly?

Yeah, we need more tests for sure.

>I guess that's missing from me in this PR, I'm trying to understand how things work, but I don't know how to verify if they work. Is running Connect My Computer the intended way to do that?

So this is how this works now:
* The client is created on the preload side, objectified and passed through the bridge.
* `createTshdClient` still exists, but it works as a compatibility layer - it is now constructed on the renderer side, not preload. Eventually we want to get rid of it.
* `connectMyComputer` service is an example of how things will look like without `createTshdClient` layer.
* All RPCs go through the new "objectified" client.
* If you want to test aborting calls, you can run the file transfer, I tested it this way.

>Take a look at my comment about awaiting the response below. https://github.com/gravitational/teleport/pull/38573#discussion_r1514709835 We could have the "objectified" client with a somewhat stricter API than the generic client provided by the library. One where maybe you cannot inspect method and requestHeaders, but one that makes the callsite use the correct way to await a response and that wraps the errors properly.

Valid point. My initial idea was to expose the client 1:1, if possible. But OTOH, we don't use `status` or `trailers` promises, so maybe it is not worth to pass them through the bridge.
Theoretically, we could objectify only `then` property. It would solve the problem with uncaught rejections, as you noticed.

Reply:
Regarding exposing a leaner client, we could expose only the uncommented properties. 
These are the ones we actually use to make calls.
The question is how to satisfy TS, probably with `as` 🥲

```ts
export function getClonedMethodsInterceptor(): RpcInterceptor {
  return {
    interceptUnary(next, method, input, options) {
      validateAbortSignalType(options.abort);

      const output = next(method, input, options);
      // UnaryCall is a class with a private field,
      // so we can't assign an object to it.
      // The workaround is satisfying only the public part of it
      // and then casting to UnaryCall.
      const clonedUnaryCall: PublicProperties<UnaryCall> = {
        // method: output.method,
        // request: output.request,
        // requestHeaders: output.requestHeaders,
        // status: clonePromiseRejection(output.status),
        // trailers: clonePromiseRejection(output.trailers),
        // headers: clonePromiseRejection(output.headers),
        // response: clonePromiseRejection(output.response),
        // then is a method on UnaryClass,
        // we have to set correct "this" before passing it further.
        then: cloneThenRejection(output.then.bind(output)),
      };
      return clonedUnaryCall as UnaryCall;
    },
    interceptClientStreaming(next, method, options) {
      validateAbortSignalType(options.abort);

      const output = next(method, options);
      const clonedClientStreamingCall: PublicProperties<ClientStreamingCall> = {
        // method: output.method,
        // requestHeaders: output.requestHeaders,
        // status: clonePromiseRejection(output.status),
        // trailers: clonePromiseRejection(output.trailers),
        // headers: clonePromiseRejection(output.headers),
        // response: clonePromiseRejection(output.response),
        requests: cloneRequests(output.requests),
        // then: cloneThenRejection(output.then.bind(output)),
      };
      return clonedClientStreamingCall as ClientStreamingCall;
    },
    interceptServerStreaming(next, method, input, options) {
      validateAbortSignalType(options.abort);

      const output = next(method, input, options);
      const clonedServerStreamingCall: PublicProperties<ServerStreamingCall> = {
        // method: output.method,
        // request: output.request,
        // requestHeaders: output.requestHeaders,
        // status: clonePromiseRejection(output.status),
        // trailers: clonePromiseRejection(output.trailers),
        // headers: clonePromiseRejection(output.headers),
        responses: cloneResponses(output.responses),
        // then: cloneThenRejection(output.then.bind(output)),
      };
      return clonedServerStreamingCall as ServerStreamingCall;
    },
    interceptDuplex(next, method, options) {
      validateAbortSignalType(options.abort);

      const output = next(method, options);
      const clonedDuplexStreamingCall: PublicProperties<DuplexStreamingCall> = {
        // method: output.method,
        // requestHeaders: output.requestHeaders,
        // status: clonePromiseRejection(output.status),
        // trailers: clonePromiseRejection(output.trailers),
        // headers: clonePromiseRejection(output.headers),
        requests: cloneRequests(output.requests),
        responses: cloneResponses(output.responses),
        // then: cloneThenRejection(output.then.bind(output)),
      };
      return clonedDuplexStreamingCall as DuplexStreamingCall;
    },
  };
}
```

Reply:
A leaner client would not have to implemented completely through interceptors alone. It could work similar to `objectifyClient`, where you call a function on the Node side that produces in a safe, "cloneable" client with a smaller API.

The function could take each method on the original client and define a new function on an object. Each function would then properly await the response and return it, along with any other thing that we need.

Reply:
I'm starting feeling that my plan of getting rid of `createTshdClient` and "objectifying" everything through the interceptors started requiring too much "magic" :/
Not being able to correctly specify types for input and output values is definitely not a good thing.

>The function could take each method on the original client and define a new function on an object.

So effectively I would have to bring back `createTshdClient` to preload :D 
But probably that's better, the end user should get a leaner API, not the API that has only a few properties working.
Also, the new `createTshdClient` could be much more concise, since ideally it would look like this:
```ts
function createTshdClient(terminalService: ITerminalServiceClient) {
  return {
    getServers: cloneUnary(terminalService.getServers),
  }
}
```
also we wouldn't need interceptors anymore, as we could do the transformation directly in `createTshdClient`:
```ts
      function cloneUnary<I extends object, O extends object>(
        s: (input: I, options?: RpcOptions) => UnaryCall<I, O>
      ): (input: I, options?: RpcOptions) => Pick<UnaryCall<I, O>, 'then'> {
        return (input, options) => {
          const original = s(input, options);
          return { then: cloneThenRejection(original.then) };
        };
      }
```
another benefit is that we could provide better types for `RpcOptions` so it would contain `ClonedAbortSignal` not the regular `AbortSignal`.

Reply:
Yeah, exactly!

> So effectively I would have to bring back `createTshdClient` to preload :D

That's not a bad thing at all. Ultimately I think the goal is to be able to retain error details and reduce the amount of boilerplate. Being able to just use abort signals without having to deal with `withAbort` is also a huge plus. Not having to do stuff in preload does not seem like a benefit by itself.